Wikipedia/Code/PushNotificationsSettingsViewController.swift

```diff
@@ -0,0 +1,281 @@
+    }
+
+    func tableView(_ tableView: UITableView, shouldHighlightRowAt indexPath: IndexPath) -> Bool {
+        return sections[indexPath.section].items[indexPath.item] as? PushNotificationsSettingsSwitchItem == nil
+    }
+
+    // MARK: - UI Actions
+
+    @objc func userDidTapSwitch(_ sender: UISwitch) {
+        let items = sections.flatMap { section in section.items }.compactMap { item in item as? PushNotificationsSettingsSwitchItem }
+        if let tappedSwitchItem = items.first(where: { item in item.tag == sender.tag }) {
+            tappedSwitchItem.action(sender.isOn)
+        }
+    }
+
+    // MARK: - Themeable
+
+    override func apply(theme: Theme) {
+        super.apply(theme: theme)
+        guard viewIfLoaded != nil else {
+            return
+        }
+        
+        view.backgroundColor = theme.colors.baseBackground
+        tableView.backgroundColor = theme.colors.baseBackground
+        tableView.reloadData()
+    }
+
+}
+
+// MARK: - System push alert/badge permissions and echo subscription/unsubscription
+
+extension PushNotificationsSettingsViewController {
+
+    fileprivate func requestPushPermissions() {
+        notificationsController.requestPermissionsIfNecessary(completionHandler: { (authorized, error) in
+            if authorized {
+                if self.notificationsController.remoteRegistrationDeviceToken == nil {
+                    // If we're still awaiting a device token, offer the user the opportunity to retry fetching one
+                    DispatchQueue.main.async {
+                        let retryAction = UIAlertAction(title: self.echoAlertFailureTryAgainActionTitle, style: .default, handler: { _ in self.requestPushPermissions() })
+                        let cancelAction = UIAlertAction(title: CommonStrings.cancelActionTitle, style: .cancel, handler: { _ in self.updateSections() })
+                        self.wmf_showAlert(title: self.echoAlertFailureTitle, message: self.echoAlertFailureMessage, actions: [retryAction, cancelAction], completion: {
+                            // Silently trigger a remote registration request to fetch a device token
+                            UIApplication.shared.registerForRemoteNotifications()
+                        })
+                    }
+                } else {
+                    // User is authorized for on device push alerts and is now awaiting subscription to Echo Notifications
+                    self.subscribeToEchoNotifications()
+                }
+            } else {
+                // User isn't authorized, just update sections to a failure state
+                self.updateSections()
+            }
+        })
+    }
+
+    fileprivate func subscribeToEchoNotifications() {
+        notificationsController.subscribeToEchoNotifications(completionHandler: { error in
+            DispatchQueue.main.async { [weak self] in
+                guard let self = self else { return }
+                self.updateSections()
```

Comment:
👍🏽 I added error alerts for both echo subscription and unsubscription. At first I thought I could get away without presenting an alert and retry operation for unsubscription, but I forgot that meant it could potentially lock the user in an unexpected state where their device would still get push messages until they retoggle the switch again.

By the way, your fix also exposed two MediaWiki API "errors" that aren't actually errors for our uses. The `echo-push-token-exists` MediaWiki code is returned during echo subscription (and parsed into a `RequestError`) if MediaWiki has already registered that device token. The `echo-push-token-not-found` MediaWiki code is returned during unsubscription (also parsed into a `RequestError`) if MediaWiki doesn't have that device token registered. In both cases, even though these are "errors", they both still represent success cases for the operations for our uses. You'll see two commits where I check to see if the error code explicitly matches these, and if so, consider the subscribe/unsubscribe request a success.

Reply:
Great catch on the non-errors! This is looking good.